Dynamic OG Image Generation Tutorial (#18384)
cloudflare/cloudflare-docs: src/content/docs/workers/tutorials/Generate Dynamic OG Images using Cloudflare Workers/index.mdx

Reviewer: Please update the name of the folder you're adding so that it's in lowercase and uses dashes (hyphens) instead of spaces. You can omit some of the words in the title.

For example:

`src/content/docs/workers/tutorials/generate-dynamic-og-images-using-workers/index.mdx`

instead of:

`src/content/docs/workers/tutorials/Generate Dynamic OG Images using Cloudflare Workers/index.mdx`

---
updated: 2024-11-23
pcx_content_type: tutorial
difficulty: Intermediate
content_type: 📝 Tutorial
title: Generate Dynamic OG Images using Cloudflare Workers
products:
  - Workers
tags:
  - OG
languages:
  - TypeScript
  - React
spotlight:
  author: Mohammed Abdulatef Al-Musaibeli
  author_bio_link: https://pwr.st
  author_bio_source: website
---

# Generate Dynamic OG Images using Cloudflare Workers

Social media thrives on compelling visuals. With Cloudflare Workers, you can effortlessly generate dynamic Open Graph (OG) images that grab attention and boost platform performance. In this tutorial, you'll learn how to create a customizable OG image generator powered by serverless edge computing. Cloudflare Workers enable you to deliver blazing-fast visuals globally, ensuring a seamless experience for your users. Let's dive in to see how you can take your OG images to the next level with Cloudflare Workers.

> What you'll accomplish:

Reviewer 2: ```suggestion
By completing this tutorial you will create an application that can:
```

    - Dynamically create stunning OG images using React and Tailwind CSS.
    - Optimize performance with built-in caching.
    - Customize visuals with flexible fonts and design options..

> By the end, you'll have a robust solution to enhance your social media presence.

Reviewer 2: ```suggestion
By the end, you will have a robust solution to enhance your social media presence.
```

GitHub repository: [Generate Dynamic OG Images using Cloudflare Workers](https://github.com/mohdlatif/og-image-generator-cloudflare-worker)

## Why Use Cloudflare Workers?

Building OG image generators can be challenging, especially when it comes to ensuring global performance, scalability, and speed. This is where Cloudflare Workers excel:

    - Global Reach: Deliver your images with ultra-low latency, no matter where your users are located.
    - Serverless Simplicity: Focus on building features, not managing infrastructure.
    - Unparalleled Performance: Process and serve requests at the edge for blazing-fast load times.

> By leveraging Cloudflare Workers, you get a serverless edge-computing environment that's not only cost-efficient but also perfectly optimized for modern web applications.

Reviewer 2: ```suggestion
By leveraging Cloudflare Workers, you get a serverless edge-computing environment that is not only cost-efficient but also perfectly optimized for modern web applications.
```

## Workflow Overview

When a user requests an OG image, the following happens:

    Request Received: A URL with parameters is sent to a Cloudflare Worker.
    Content Processed: The Worker extracts text, style, and font configurations from the URL.
    Font Loaded: Fonts are retrieved using one of four methods, like Google Fonts or local files.
    Image Generated: The image is built using React components styled with Tailwind CSS.
    Response Cached: The final image is returned and cached for future requests.

> Here's how the process flows:

Reviewer 2: ```suggestion
Here is how the process flows:
```

    "strict": true, // Enable all strict type checking options
    "skipLibCheck": true, // Skip type checking of declaration files
    "noFallthroughCasesInSwitch": true,
    // Some stricter flags (disabled by default)
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "noPropertyAccessFromIndexSignature": false
  },
  "baseUrl": "./", // Base directory for resolving non-relative module names
  "paths": {
    "@/*": [ // Path alias configuration
      "./src/*"
    ]
  }
}

```

Key TypeScript configuration features:

- React JSX support with `jsx: "react-jsx"`
- ES modules configuration with `module: "ESNext"`
- Cloudflare Workers types integration
- Path aliases for cleaner imports
- Strict type checking enabled
- Modern JavaScript features support

### Configure Wrangler for Runtime Compatibility and Static Assets

Before starting, ensure your `wrangler.toml` includes these essential configurations:

```toml
compatibility_date = "2024-11-06"
compatibility_flags = [ "nodejs_compat_v2" ]
assets = { directory = "public",  binding = "ASSETS" }
minify = true
[build]
watch_dir = "public"
```

> The `nodejs_compat_v2` flag enables runtime compatibility features required by the OG image generation library, even when using Bun. While we're using Bun as our development runtime, this flag ensures all necessary APIs are available in the Workers environment. The `assets` configuration maps your Worker's public directory, allowing direct access to static files like fonts, images, and favicons through URL paths (for example, `/fonts/Inter.ttf`, `/images/logo.png`).

Reviewer 2: ```suggestion
The `nodejs_compat_v2` flag enables runtime compatibility features required by the OG image generation library, even when using Bun. While using Bun as our development runtime, this flag ensures all necessary APIs are available in the Workers environment. The `assets` configuration maps your Worker's public directory, allowing direct access to static files like fonts, images, and favicons through URL paths (for example, `/fonts/Inter.ttf`, `/images/logo.png`).
```

- Automatically detects image content type
- Converts images to base64 for use in OG images
- Provides fallback handling if image loading fails

Use it in your templates like this:

```typescript
const logoImage = await loadImage(c, '/images/your-logo.png');
```

:::note
Make sure your images are stored in the `public/images` directory and referenced with paths starting with `/images/`.
:::

## 6. Implement the OG image generator

Create the main image generation handler:

```typescript

---

filename: src/og.tsx

---

import { Hono } from 'hono';

import { ImageResponse } from '@cloudflare/pages-plugin-vercel-og/api';

const app = new Hono();

export default app.get('/', async (c) => {

  const { mainText, description, footerText } = c.req.query();

  // Implementation details

});

```

## 7. Add visual styles

The OG image generator includes 4 distinct styles that can be selected via the `style` query parameter. The style selection is handled through a simple query parameter in the URL:

```typescript
style = 1; // Default professional style
style = 2; // Eco-tech theme
style = 3; // Corporate brand style
style = 4; // GitHub profile style
```

> If no style parameter is provided or an invalid value is used, the generator defaults to Style1. Here's how to use each style:

Reviewer 2: ```suggestion
If no style parameter is provided or an invalid value is used, the generator defaults to Style1. Here is how to use each style:
```

```url
/og?style=4
```

Features:

- Minimal design
- GitHub avatar integration
- Perfect for developer profiles

:::tip
You can combine any style with other parameters like `mainText`, `description`, and `footerText` to customize the output further.
:::

The style selection is implemented using a ternary chain in the code:

```typescript

---

filename: src/index.ts

---

const SocialCardTemplate =
    c.req.query('style') === '2'
        ? Style2()
        : c.req.query('style') === '3'
        ? Style3()
        : c.req.query('style') === '4'
        ? Style4()
        : Style1();
```

This implementation allows for easy addition of new styles in the future by simply adding new conditions to the chain and corresponding style components.

```typescript

---

filename: src/og.tsx

---

function Style1() {

  return (

    <div tw="flex flex-col w-full h-full p-12 bg-gradient-to-br from-blue-900 to-indigo-700">

      {/* Style implementation */}

    </div>

  );

}

```

## 8. Configure caching

Enable caching to:

- Reduce computation costs
- Improve response times
- Decrease origin server load
- Provide consistent performance

> Here's how to implement caching with customizable durations:

Reviewer 2: ```suggestion
Add the following code to implement caching with customizable durations:
```

```typescript

---

filename: src/index.ts

---

import { Hono } from 'hono';

import { cache } from 'hono/cache';

const app = new Hono()

  .use('*', cache({

    cacheName: async (c) => {

      const url = new URL(c.req.url);

      return `${c.req.method} ${url.pathname}${url.searchParams}`;

    },

    cacheControl: 'max-age=86400'

  }))

  .route('og', og);

```

:::caution

Make sure to configure appropriate cache durations based on your application's needs. The example uses a 24-hour cache duration.

:::

## Setting Up TypeScript Type Definitions

> Let's create our type definitions to ensure smooth TypeScript integration. Create a new file `src/type.d.ts`:

Reviewer 2: ```suggestion
Create a new file `src/type.d.ts`, then add the following code to the new file to create type definitions to ensure smooth TypeScript integration:
```

```typescript

---

filename: src/type.d.ts

---

interface CacheStorage {
	default: Cache;
}

declare module '@cloudflare/pages-plugin-vercel-og/api' {
	import { ImageResponse as VercelImageResponse } from '@vercel/og';
	export declare class ImageResponse extends Response {
		constructor(...args: ConstructorParameters<typeof VercelImageResponse>);
	}
}

declare module '*.woff2' {
	const content: ArrayBuffer;
	export default content;
}

declare module '*.woff' {
	const content: ArrayBuffer;
	export default content;
}

declare module '*.ttf' {
	const content: ArrayBuffer;
	export default content;
}

```

These type definitions are crucial for our project:

- They enable TypeScript to understand Cloudflare's cache storage
- Add proper typing for the Vercel OG image response
- Allow direct importing of font files (`.woff2`, `.woff`, `.ttf`)

> With these definitions in place, you'll get full TypeScript support and autocompletion throughout your project. This makes development smoother and helps catch potential issues early.

Reviewer 2: ```suggestion
With these definitions in place, you will get full TypeScript support and autocompletion throughout your project. This makes development smoother and helps catch potential issues early.
```

## 9. Deploy your Worker

Deploy the application to Cloudflare Workers:

```bash

bun run deploy

```

## Usage examples

Generate OG images by making GET requests:

```url
https://your-worker.workers.dev/og?mainText=Hello%20World&description=A%20dynamic%20OG%20image&style=1
```

You can customize the image by adjusting query parameters:

- `mainText`: Main heading

- `description`: Detailed description

- `footerText`: Footer content

- `style`: Visual style (1-4)

## Conclusion

> Congratulations! You've successfully harnessed the power of Cloudflare Workers to build a dynamic OG image generator. By leveraging the global edge network and serverless architecture, your application now delivers high-performance visuals with ease. Whether you're scaling for millions of users or iterating on design tweaks, Cloudflare Workers ensure your images are always fast, reliable, and stunning.

Reviewer 2: ```suggestion
Congratulations! You have successfully harnessed the power of Cloudflare Workers to build a dynamic OG image generator. By leveraging the global edge network and serverless architecture, your application now delivers high-performance visuals with ease. Whether you are scaling for millions of users or iterating on design tweaks, Cloudflare Workers ensure your images are always fast, reliable, and stunning.
```

This solution provides:

- Fast generation times through edge computing
- Multiple font loading options
- Pre-designed visual styles
- Built-in caching support

The complete source code is available on [GitHub](https://github.com/mohdlatif/og-image-generator-cloudflare-worker).

You have successfully created a dynamic OG image generator that runs at the edge. The application can now generate social media cards with:

## Related resources

> - [Cloudflare Workers documentation](https://developers.cloudflare.com/workers/)

Reviewer 2: ```suggestion
- [Cloudflare Workers documentation](/workers/)
```

- [Hono framework](https://hono.dev/)

- [Vercel OG Image Generation](https://vercel.com/docs/functions/og-image-generation)

> - [Cloudflare page plugin vercel/og](https://developers.cloudflare.com/pages/functions/plugins/vercel-og/)

Reviewer 2: ```suggestion
- [Cloudflare page plugin vercel/og](/pages/functions/plugins/vercel-og/)
```
